internal/app/client/conformance_client.go

```diff
@@ -246,91 +252,88 @@ func (i *invoker) bidiStream(
-					// to a proto Error
+					// to a proto Error. If the error was an EOF, that just means
+					// reads are done.
 					protoErr = app.ConvertErrorToProtoError(err)
 				}
 				// Reads are done either because we received an error or an EOF
 				// In either case, break the outer loop
 				break
 			}
 			// If the call was successful, get the returned payloads
-			payloads = append(payloads, msg.Payload)
+			result.Payloads = append(result.Payloads, msg.Payload)
 		}
 	}
 
-	// If protoErr is nil, then the send loop terminated successfully
-	if protoErr == nil {
-		// Close the send side of the stream
-		if err := stream.CloseRequest(); err != nil {
-			protoErr = app.ConvertErrorToProtoError(err)
-		}
+	// If we received an error in any of the send logic or full-duplex reads, then exit
+	if protoErr != nil {
+		result.Error = protoErr
+		return result, nil
+	}
+
+	// Sends are done, close the send side of the stream
+	if err := stream.CloseRequest(); err != nil {
+		result.Error = app.ConvertErrorToProtoError(err)
+		return result, nil
 	}
 
-	// If there were no errors encountered yet, then receive any additional responses
-	if protoErr == nil {
-		for {
-			if err := ctx.Err(); err != nil {
-				// If an error was returned, convert it to a proto Error
+	// Receive any remaining responses
+	for {
+		if err := ctx.Err(); err != nil {
+			// If an error was returned, convert it to a proto Error
+			protoErr = app.ConvertErrorToProtoError(err)
+			break
+		}
+		msg, err := stream.Receive()
+		if err != nil {
+			if !errors.Is(err, io.EOF) {
+				// If an error was returned that is not an EOF, convert it
+				// to a proto Error. If the error was an EOF, that just means
+				// reads are done.
 				protoErr = app.ConvertErrorToProtoError(err)
-				break
-			}
-			msg, err := stream.Receive()
-			if err != nil {
-				// TODO - Should we check whether err is io.EOF? Because if not,
-				// then we have an error that should be recorded probably?
-				// OR will this be unwrapped via the CloseResponse call below?
-				// Reads are done break the outer loop
-				break
 			}
-			// If the call was successful, get the returned payloads
-			// and the headers and trailers
-			payloads = append(payloads, msg.Payload)
+			break
 		}
+		// If the call was successful, save the payloads
+		result.Payloads = append(result.Payloads, msg.Payload)
+	}
 
-		// TODO - Related to the above TODO, will this unwrap the error from the
-		// receive call?
-		if err := stream.CloseResponse(); err != nil {
-			protoErr = app.ConvertErrorToProtoError(err)
-		}
+	if protoErr != nil {
+		result.Error = protoErr
+		return result, nil
 	}
 
-	// Read headers and trailers from the stream
-	headers = app.ConvertToProtoHeader(stream.ResponseHeader())
-	trailers = app.ConvertToProtoHeader(stream.ResponseTrailer())
+	// TODO - Does this unwrap the error from the receive call or is this
+	// a separate error that needs handled?
```

Comment:
You always return any error from a receive call, so if this returns an error, it will be something new I think. Unclear though since a non-nil error from receive effectively means "end of stream", so there may be nothing else that can happen here.

In any event, I guess I'm saying that the code you have here looks good to me.